cpp/src/sampling/detail/sampling_utils_impl.cuh

```diff
@@ -0,0 +1,886 @@
+                 global_adjacency_list_offsets.end(),
+                 edge_t{0});
+  }
+
+  return global_adjacency_list_offsets;
+}
+
+template <typename GraphViewType>
+std::tuple<rmm::device_uvector<typename GraphViewType::edge_type>,
+           rmm::device_uvector<typename GraphViewType::edge_type>>
+get_global_degree_information(raft::handle_t const& handle, GraphViewType const& graph_view)
+{
+  using edge_t       = typename GraphViewType::edge_type;
+  auto local_degrees = compute_local_major_degrees(handle, graph_view);
+
+  if constexpr (GraphViewType::is_multi_gpu) {
+    auto& col_comm      = handle.get_subcomm(cugraph::partition_2d::key_naming_t().col_name());
+    auto const col_size = col_comm.get_size();
+    auto const col_rank = col_comm.get_rank();
+    auto& row_comm      = handle.get_subcomm(cugraph::partition_2d::key_naming_t().row_name());
+    auto const row_size = row_comm.get_size();
+
+    auto& comm           = handle.get_comms();
+    auto const comm_size = comm.get_size();
+    auto const comm_rank = comm.get_rank();
+
+    rmm::device_uvector<edge_t> temp_input(local_degrees.size(), handle.get_stream());
+    raft::update_device(
+      temp_input.data(), local_degrees.data(), local_degrees.size(), handle.get_stream());
+
+    rmm::device_uvector<edge_t> recv_data(local_degrees.size(), handle.get_stream());
+    if (col_rank == 0) {
+      thrust::fill(handle.get_thrust_policy(), recv_data.begin(), recv_data.end(), edge_t{0});
+    }
+    for (int i = 0; i < col_size - 1; ++i) {
+      if (col_rank == i) {
+        comm.device_send(
+          temp_input.begin(), temp_input.size(), comm_rank + row_size, handle.get_stream());
+      }
+      if (col_rank == i + 1) {
+        comm.device_recv(
+          recv_data.begin(), recv_data.size(), comm_rank - row_size, handle.get_stream());
+        thrust::transform(handle.get_thrust_policy(),
+                          temp_input.begin(),
+                          temp_input.end(),
+                          recv_data.begin(),
+                          temp_input.begin(),
+                          thrust::plus<edge_t>());
+      }
+      col_comm.barrier();
+    }
+    // Get global degrees
+    device_bcast(col_comm,
+                 temp_input.begin(),
+                 temp_input.begin(),
+                 temp_input.size(),
+                 col_size - 1,
+                 handle.get_stream());
+
+    return std::make_tuple(std::move(recv_data), std::move(temp_input));
+  } else {
+    rmm::device_uvector<edge_t> temp(local_degrees.size(), handle.get_stream());
+    thrust::fill_n(handle.get_thrust_policy(), temp.begin(), local_degrees.size(), edge_t{0});
+    return std::make_tuple(std::move(temp), std::move(local_degrees));
+  }
+}
+
+template <typename vertex_t>
+rmm::device_uvector<vertex_t> gather_active_majors(raft::handle_t const& handle,
+                                                   rmm::device_uvector<vertex_t>&& d_in)
```

Comment:
OK, this is using allgatherv, so this function should better be renamed to "allgather_active_majors".

Reply:
Fixed in next push